File: neetcode150/medium/347-Top-K-Frequent-Elements.py
from typing import List

class Solution:
  # Intital solution O(k log n)
  def topKFrequent(self, nums: List[int], k: int) -> List[int]:
    d = {} 
    for n in nums:
      if n in d:
        d[n] += 1
      else:
        d[n] = 1
        
    s = []
    m = int()
    i = int()
    for v in range(k):
      m = int()
      for e in d:
        if d[e] > m:
          m = max(m, d[e])
          i = e
      del d[i]
      s.append(i) 
    return s

  # Optimal solution O(n) using bucket sort
  def topKFrequent(self, nums: List[int], k: int) -> List[int]:
      h = {}
      freq = [[] for i in range(len(nums) + 1)]

      for n in nums:
          h[n] = h.get(n, 0) + 1
        
      for n, c in h.items():
          freq[c].append(n)

      res = []
      for i in range(len(freq) - 1, 0, -1):
          bucket = freq[i]
          while bucket:
              res.append(bucket.pop())
              k -= 1
              if k == 0:
                  return res

File: neetcode150/medium/cpp/347-Top-K-Frequent-Elements.cpp
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int, int> h;
        for (int n : nums) {
            h[n] += 1;
        }

        vector<vector<int> > b(nums.size()+1);
        for (auto& [n, v] : h) {
            b[v].push_back(n);
        }

        vector<int> result;
        for (int i = b.size()-1; i > 0; --i) {
            vector<int> bucket = b[i];
            while (!bucket.empty()) {
                result.push_back(bucket.back());
                bucket.pop_back();
                --k;
                if (k == 0) {
                    return result;
                }
            }
        }
        
        return result;
    }
};